src/main/java/io/gravitee/policy/jwt/jwk/source/JWKSUrlJWKSourceResolver.java

```diff
@@ -0,0 +1,133 @@
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.policy.jwt.jwk.source;
+
+import com.nimbusds.jose.KeySourceException;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.JWKSelector;
+import com.nimbusds.jose.jwk.JWKSet;
+import com.nimbusds.jose.jwk.source.ImmutableJWKSet;
+import com.nimbusds.jose.jwk.source.JWKSource;
+import com.nimbusds.jose.proc.SecurityContext;
+import com.nimbusds.jose.util.Resource;
+import io.reactivex.Completable;
+import io.reactivex.Maybe;
+import io.reactivex.schedulers.Schedulers;
+import java.net.MalformedURLException;
+import java.text.ParseException;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Specific implementation of {@link JWKSource} allowing retrieve the JWKS from an url and triggering refresh every {@link #CACHE_DURATION} in background.
+ * If an error occurred during background refresh, previous version is kept to limit negative impacts.
+ *
+ * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class JWKSUrlJWKSourceResolver<C extends SecurityContext> implements JWKSource<C> {
+
+    private static final Logger log = LoggerFactory.getLogger(JWKSUrlJWKSourceResolver.class);
+
+    private final String jwksUrl;
+    private final Duration refreshInterval;
+    private final AtomicBoolean refreshing;
+    private final ResourceRetriever resourceRetriever;
+
+    /**
+     * Global cache allowing to share jwk source across the whole platform and avoid re-loading the same JWKS url multiple times.
+     */
+    static final ConcurrentHashMap<String, CachedJWKSource<SecurityContext>> cache = new ConcurrentHashMap<>();
+
+    public JWKSUrlJWKSourceResolver(String jwksUrl, ResourceRetriever resourceRetriever, Duration refreshInterval) {
+        this.jwksUrl = jwksUrl;
+        this.refreshInterval = refreshInterval;
+        this.refreshing = new AtomicBoolean(false);
+        this.resourceRetriever = resourceRetriever;
+    }
+
+    @Override
+    public List<JWK> get(JWKSelector jwkSelector, C context) throws KeySourceException {
+        CachedJWKSource<SecurityContext> cachedJWKSource = cache.get(jwksUrl);
+
+        if (cachedJWKSource != null) {
+            if (isCacheExpired(cachedJWKSource)) {
+                backgroundRefresh();
```

Comment:
The idea is to trigger a refresh of the JWKS while still using the previous version in cache. That way, the current request is not impacted by the latency induced by the jwks retrieval and it will benefit to the next requests.
Makes sense to you ?

Reply:
ok I got it, but that means for the current request you will use a expired jwks from the cache. Is it an issue?